ps3.py

```diff
@@ -0,0 +1,236 @@
+        else:
+            return False
+    elif word in word_list:
+        for letter in word:
+            if letter not in hand or word.count(letter) > hand.get(letter, 0):
+                return False
+        else:
+            return True
+    else:
+        return False
+
+
+def calculate_handlen(hand):
+    """
+    Returns the length (number of letters) in the current hand.
+    """
+    count_letter = 0
+    for letter in hand.keys():
+        count_letter += hand[letter]
+    return count_letter
+
+
+def play_hand(hand, word_list, count_points):
+    """
+    The hand finishes when there are no more unused letters.
+    The user can also finish playing the hand by inputing two
+    exclamation points (the string '!!') instead of a word.
+    """
+    while calculate_handlen(hand) != 0:
+        print("Current Hand:", end=" ")
+        display_hand(hand)
+        user_word = input("Enter word, or “!!” to indicate that you are finished:")
+        if is_valid_word(user_word, hand, word_list):
+            score_word = get_word_score(user_word, calculate_handlen(hand))
+            count_points += score_word
+            print(f"{user_word} earned {str(score_word)}."
+                  f" Total: {count_points} points")
+            print(" ")
+        elif user_word == "!!":
+            return count_points
+        else:
+            print("This is not valid word. Please choose another word")
+        hand = update_hand(hand, user_word)
+    else:
+        return count_points
+
+
+def substitute_hand(hand, letter):
+    """
+    Allow the user to replace all copies of one letter in the hand (chosen by user)
+    with a new letter chosen from the VOWELS and CONSONANTS at random. The new letter
+    should be different from user's choice, and should not be any of the letters
+    already in the hand.
+
+    If user provide a letter not in the hand, the hand should be the same.
+
+    Has no side effects: does not mutate hand.
+    """
+    if letter in hand:
+        new_hand = hand.copy()
+        while True:
+            new_letter = random.choice(VOWELS and CONSONANTS)
+            if new_letter not in hand:
+                new_hand[new_letter] = new_hand.pop(letter)
+                return new_hand
+    return hand
+
+
+def play_game(word_list):
+    """
+    Allow the user to play a series of hands
+
+    * Asks the user to input a total number of hands
+
+    * Accumulates the score for each hand into a total score for the
+      entire series
+
+    * Call functions play_hand, it returns a score of hand
+
+    * Returns the total score for the series of hands
+    """
+    count_final_points = 0
+    number_of_hands = int(input("Enter total number of hands:"))
+
+    while number_of_hands != 0:
+        hand = deal_hand(HAND_SIZE)
+        print("Current hand:", end=" ")
+        display_hand(hand)
+        ask_user = str(input("Would you like to substitute a letter?"))
+        if ask_user == "yes":
```

Comment:
User can substitute only once.
If user have replayed the hand, he can't substitute a letter.